src/Microsoft.AspNetCore.WebHooks.Api/Controllers/WebHookRegistrationsController.cs

```diff
@@ -0,0 +1,315 @@
+            List<string> invalidFilters = new List<string>();
+            foreach (string filter in webHook.Filters)
+            {
+                WebHookFilter hookFilter;
+                if (filters.TryGetValue(filter, out hookFilter))
+                {
+                    normalizedFilters.Add(hookFilter.Name);
+                }
+                else
+                {
+                    invalidFilters.Add(filter);
+                }
+            }
+
+            if (invalidFilters.Count > 0)
+            {
+                string invalidFiltersMsg = string.Join(", ", invalidFilters);
+                string link = Url.Link(WebHookRouteNames.FiltersGetAction, values: null);
+                string msg = string.Format(CultureInfo.CurrentCulture, Properties.CustomApiResources.RegistrationController_InvalidFilters, invalidFiltersMsg, link);
+                _logger.LogInformation(msg);
+
+                throw new NotImplementedException(msg);
+            }
+            else
+            {
+                webHook.Filters.Clear();
+                foreach (string filter in normalizedFilters)
+                {
+                    webHook.Filters.Add(filter);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensures that the provided <paramref name="webHook"/> has a reachable Web Hook URI.
+        /// </summary>
+        protected virtual async Task VerifyWebHook(WebHook webHook)
+        {
+            if (webHook == null)
+            {
+                throw new ArgumentNullException("webHook");
+            }
+
+            try
+            {
+                await _manager.VerifyWebHookAsync(webHook);
+            }
+            catch (Exception ex)
+            {
+                throw new Exception(ex.Message, ex);
+            }
+        }
+
+        /// <summary>
+        /// Gets the user ID for this request.
+        /// </summary>
+        protected virtual async Task<string> GetUserId()
+        {
+            try
+            {
+                string id = await _user.GetUserIdAsync(User);
+                return id;
+            }
+            catch (Exception ex)
+            {
+                throw new Exception(ex.Message, ex);
+            }
+        }
+
+        /// <summary>
+        /// Creates an <see cref="IHttpActionResult"/> based on the provided <paramref name="result"/>.
+        /// </summary>
+        /// <param name="result">The result to use when creating the <see cref="IHttpActionResult"/>.</param>
+        /// <returns>An initialized <see cref="IHttpActionResult"/>.</returns>
+        protected IActionResult CreateHttpResult(StoreResult result)
+        {
+            switch (result)
+            {
+                case StoreResult.Success:
+                    return Ok();
+
+                case StoreResult.Conflict:
+                    return new HttpStatusCodeResult((int)System.Net.HttpStatusCode.Conflict);

```

Comment:
BTW you can use these const int's instead: [https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/StatusCodes.cs](https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/StatusCodes.cs)

Then there's no need to cast.